# 4.7 Deriving New Variables Exercise

## Exercise practice follow along

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Create path

path = r"C:\Users\leobs\OneDrive\Data Program\Achievement 4_Instacart\Instacart Basket Analysis"

In [4]:
# Import dataframe

ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [5]:
# Check for import

ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [7]:
# Create subset df for practice

df = ords_prods_merged[:1000000]

In [8]:
df.shape

(1000000, 14)

## Using user-defined function on subset

In [9]:
# Define a function

def price_label(row):
    if row['prices'] <=5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif row['prices'] > 15:
        return 'High range'
    else: return 'Not enough data'

In [11]:
# Apply the function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\leobs\AppData\Local\Temp\ipykernel_28632\93659844.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [12]:
# Check for new column

df.value_counts()

# Note no High-range product labels, but currently only looking at subset of df

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  days_since_last_order  product_id  add_to_cart_order  reordered  product_name                             aisle_id  department_id  prices  _merge  price_range      
5         156122   42            6                  16                 9.0                    13176       1                  1          Bag of Organic Bananas                   24        4              10.3    both    Mid-range product    1
2281487   108843   5             1                  13                 13.0                   47766       28                 1          Organic Avocado                          24        4              6.3     both    Mid-range product    1
2281494   123967   53            0                  14                 6.0                    13176       3                  1          Bag of Organic Bananas                   24        4              10.3    both    Mid-range product    1
2281495   147802   43            1       

In [13]:
df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_range  
0         Soda        77              7     9.0   both  Mid-range product  
1         Soda        77              7     9.0   both  Mid-range product  
2         Soda        77              7     9.0   both  Mid-range product  
3         Soda        77              7     9.0   both  Mid-range product  
4         Soda        77              7     9.0   both  Mid-range product

In [14]:
# Check for max

df['prices'].max()

# Helps check to make sure no errors in calculations due to lack of High-range product labels

14.8

## Using loc() function on subset

Define conditions for loc()

In [16]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [17]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [18]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [20]:
# Check product counts for each label

df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [21]:
# Check for same results as user-created function

df.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_range  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

     price_range_loc  
0  Mid-range product  
1  Mid-range product  
2  Mid-range product  
3  Mid-range product  
4  Mid-range product

## Run loc() function on ords_prods_merge full df

In [22]:
ords_prods_merged.loc[ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [26]:
ords_prods_merged.loc[(ords_prods_merged['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [28]:
ords_prods_merged.loc[ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [29]:
# Check header

ords_prods_merged.head()

# Successful

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  
0         Soda        77              7     9.0   both  Mid-range product  
1         Soda        77              7     9.0   both  Mid-range product  
2         Soda        77              7     9.0   both  Mid-range product  
3         Soda        77              7     9.0   both  Mid-range product  
4         Soda        77              7     9.0   both  Mid-range product

In [32]:
ords_prods_merged['price_range_loc'].value_counts(dropna = False)

NaN                   21104410
Low-range product     10126321
Mid-range product       756450
High-range product      417678
Name: price_range_loc, dtype: int64

## If-statements with for loops

## Use for-loop to create a column that summarizes how busy each day of the week is

In [38]:
# Identify day of the week with most orders

ords_prods_merged['order_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

Value 0 means Saturday, therefore Saturday has highest number of orders and busiest day
Wednesday (value 4) has lowest frequency and is slowest day of the week

In [39]:
# Create for-loop for busiest day

result = []

for value in ords_prods_merged['order_day_of_week']:
    if value == 0:
        result.append('Busiest day')
    elif value == 4:
        result.append('Least busy')
    else:
        result.append('Regularly busy')

In [40]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [41]:
# Combine result list with ords_prods_merged dataframe in new column, 'busiest_day'

ords_prods_merged['busiest_day'] = result

In [42]:
# Print frequency to cross check

ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

In [43]:
# Cross check with order_day_of_week

ords_prods_merged['order_day_of_week'].value_counts(dropna = False)

# Shows accurate match

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

In [44]:
# Check updated dataframe for new column

ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

Saving changes due to pause in exercise

In [45]:
# Export df

ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged.pkl'))

# Task 4.7 Creating new columns in ords_prods_merged dataframe

## Step 1 price_label and busiest_day columns

Note: already completed these steps, however need to rename price_range_loc to price_label

In [7]:
# Change name of price_range_loc to price_label

ords_prods_merged.rename(columns = {'price_range_loc' : 'price_label'}, inplace = True)

In [8]:
# Check rename

ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_label  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3      Least busy  
4      Least busy

## Step 2 Create 'busiest days' column

In [9]:
# Locate two busiests and two slowest days of the week

ords_prods_merged['order_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

0 and 1 are busiests, representing Saturday and Sunday
3 and 4 are slowest, representing Wednesday and Thursday
2, 5, 6 are average, representing Monday, Thursday, and Friday

In [11]:
# Create a for-loop for requested busiest days column

result_busiest_days = []

for value in ords_prods_merged['order_day_of_week']:
    if value == 0 or value == 1:
        result_busiest_days.append('Busiest days')
    if value == 3 or value == 4:
        result_busiest_days.append('Slowest days')
    else:
        result_busiest_days.append('Regularly busy days')

In [12]:
# Print result

result_busiest_days

['Regularly busy days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Slowest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Slowest days',
 'Regularly busy days',
 'Slowest

In [13]:
# Merge with dataframe as a new column

ords_prods_merged['busiest_days'] = result_busiest_days

ValueError: Length of values (44269271) does not match length of index (32404859)

In [17]:
# Correction with list, used if instead of elif, update for-loop

result_busiest_days = []

for value in ords_prods_merged['order_day_of_week']:
    if value == 0 or value == 1:
        result_busiest_days.append('Busiest days')
    elif value == 3 or value == 4:
        result_busiest_days.append('Slowest days')
    else:
        result_busiest_days.append('Regularly busy days')

In [18]:
# Print list

result_busiest_days

['Regularly busy days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Regularly busy days',
 'Regularly busy days',
 'Regularly busy days',
 'Slowest days',
 'Regularly busy days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Slowest days',
 'Slowest days',
 'Regularly busy days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regul

In [19]:
# Merge with dataframe as a new column

ords_prods_merged['busiest_days'] = result_busiest_days

## Step 3 check values of new column

In [20]:
# Check column

ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_label  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days  
0  Regularly busy  Regularly busy days  
1  Regularly busy         Slowest days  
2  Regularly busy         Slowest days  
3      Least busy         Slowest days  
4      Least busy         Slowest days

In [21]:
# Check value counts

ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy days    12916111
Busiest days           11864412
Slowest days            7624336
Name: busiest_days, dtype: int64

In [22]:
# Check value counts

ords_prods_merged['order_day_of_week'].value_counts(dropna = False)m

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: order_day_of_week, dtype: int64

Accurate match in value counts and successfully added new column showing busiest, regular, and slowest days of the week

Each grouping matches the sum of the relevant columns

## Step 4 identify busiest hours and create new column

In [23]:
# Identify busiest hours

ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

Observations:
Client wants three categories
Days 10, 11, 14, 15, 13, 12, 16, 9, and 17 all have above 200000 orders and will be grouped together as most
Days 8, 18, 19, 20, 7, 21, 22, 23, 6, 0, and 1 all have between 100000 and 200000 orders and will be grouped as average
Days 5, 2, 4, and 3 all have less than 100000 orders and will be grouped as fewest

Note, using actual number of orders rather than splitting the day equally into 3 categories by same number of hours do to the pattern in number of orders across hours of the day

In [25]:
# create subset to run summary stats to check grouping choice

days_counts = ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

In [26]:
days_counts

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [27]:
days_counts.describe()

# keeping choice to have 3 groups based on counts, but rather equal ranges of 0 to 100000, 100001 to 20000, and 200001 to 300000

count    2.400000e+01
mean     1.350202e+06
std      1.078369e+06
min      5.128100e+04
25%      2.725620e+05
50%      1.117230e+06
75%      2.556034e+06
max      2.761760e+06
Name: order_hour_of_day, dtype: float64

In [28]:
# Create for-loop for busiest hours

hours = []

for value in ords_prods_merged['order_hour_of_day']:
    if value in [10, 11, 14, 15, 13, 12, 16, 9, 17]:
        hours.append('Most orders')
    elif value in [5, 2, 4, 3]:
        hours.append('Fewest orders')
    else:
        hours.append('Average orders')

In [29]:
hours

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders

In [30]:
# Merge hours list with dataframe as new column

ords_prods_merged['Busiest_period_of_day'] = hours

## Step 5 check frequency of new column

In [31]:
# Check column headers

ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_label  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days Busiest_period_of_day  
0  Regularly busy  Regularly busy days        Average orders  
1  Regularly busy         Slowest days        Average orders  
2  Regularly busy         Slowest days           Most orders  
3      Least busy         Slowest days        Average orders  
4      Least busy         Slowest days           Most orders

In [38]:
# change capitalization

ords_prods_merged.rename(columns = {'Busiest_period_of_day' : 'busiest_period_of_day'}, inplace = True)

#unsure why output is complete df rather than just changing the column header?- forgot inplace = true, reran with it

In [39]:
# Check for rename

ords_prods_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_label  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy days        Average orders  
1  Regularly busy         Slowest days        Average orders  
2  Regularly busy         Slowest days           Most orders  
3      Least busy         Slowest days        Average orders  
4      Least busy         Slowest days           Most orders

In [40]:
# Check value counts

ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       23205725
Average orders     8937275
Fewest orders       261859
Name: busiest_period_of_day, dtype: int64

Value counts match the 32404859 figure

## Step 6 and 7 clean and export dataframe

In [41]:
# Export datagrame as pickle


ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged.pkl'))